src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+pub trait Message {
+    fn get_size(&self) -> i32;
+}
+
+pub trait FromByte: Default {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors>;
+
+    fn decode_new(buf: &mut Buf) -> Result<Self, DecodeErrors> {
+        let mut temp = Self::default();
+        match temp.decode(buf) {
+            Ok(_) => Ok(temp),
+            Err(e) => Err(e),
```

Comment:
Rather than matching `Err(e) => Err(e)`, you could also say `temp.decode(buf)?` above. The `?` operator would unwrap the `Ok` variant if the expression is `Ok`, or cause your function return `Err` immediately if the expression is an error.

Reply:
Got it. Thx.